torch/csrc/jit/passes/frozen_conv_add_relu_fusion.cpp

```diff
@@ -117,10 +119,92 @@ void fuseFrozenConvAddReluImpl(std::shared_ptr<Graph>& graph) {
+
+  auto conv_relu_rstring_dense = CodeTemplate(R"(
+    graph(%input, %weight, %bias, %stride:int[], %padding:int[], %dilation:int[], %groups:int):
+      %x = aten::${conv}(%input, %weight, %bias, %stride, %padding, %dilation, %groups)
+      %res = aten::${relu}(%x)
+      return (%res))");
+
+  auto conv_relu_rstring_mkldnn = CodeTemplate(R"(
+    graph(%input, %weight, %bias, %stride:int[], %padding:int[], %dilation:int[], %groups:int):
+      %x = prim::mkldnn_convolution(%input, %weight, %bias, %stride, %padding, %dilation, %groups)
+      %res = aten::${relu}(%x)
+      return (%res))");
+
+  std::string conv_relu_fused = R"(
+    graph(%input, %weight, %bias, %stride:int[], %padding:int[], %dilation:int[], %groups:int):
+        %res = aten::convolution_relu(%input, %weight, %bias, %stride, %padding, %dilation, %groups)
+        return (%res))";
+
+  // dense cpu tensor input path.
+  for (const auto& conv : conv_operators) {
+    for (const auto& relu : relu_operators) {
+      TemplateEnv env;
+      env.s("conv", conv);
+      env.s("relu", relu);
+      rewriter_dense_input.RegisterRewritePattern(
+          conv_relu_rstring_dense.format(env), conv_relu_fused);
+    }
+  }
+
+  // mkldnn tensor input path.
+  for (const auto& relu : relu_operators) {
+    TemplateEnv env;
+    env.s("relu", relu);
+    rewriter_mkldnn_input.RegisterRewritePattern(
+        conv_relu_rstring_mkldnn.format(env), conv_relu_fused);
+  }
+
+  auto filter = [](const Match& match,
+                   const std::unordered_map<std::string, Value*>& vmap) {
+    auto weight = toIValue(match.values_map.at(vmap.at("weight")));
+    if (!weight.has_value() || !weight.value().isTensor()) {
+      return false;
+    }
+
+    const at::Tensor& weight_t = weight.value().toTensor();
+    if (!(weight_t.device().is_cpu() && weight_t.scalar_type() == at::kFloat) ||
+        !weight_t.is_contiguous()) {
+      return false;
+    }
+
+    // bias is optional
+    if (vmap.find("bias") != vmap.end()) {
+      auto bias = toIValue(match.values_map.at(vmap.at("bias")));
+      if (bias.has_value() && bias.value().isTensor()) {
+        const at::Tensor& bias_t = bias.value().toTensor();
+        if (bias_t.dtype() != weight_t.dtype() || bias_t.ndimension() != 1 ||
+            bias_t.size(0) != weight_t.size(0) || !bias_t.is_contiguous() ||
+            !bias_t.device().is_cpu()) {
+          return false;
+        }
+      }
+    }
+
+    return true;
+  };
+
+  // Convert _convolution and in-place operators for simpler replacement pattern
+  // matching
+  graph_rewrite_helper::replaceConvolutionWithAtenConv(graph);
+  rewriter_dense_input.runOnGraph(graph, filter);
+  rewriter_mkldnn_input.runOnGraph(graph);
+#endif
+}
 } // namespace
 
 void FuseFrozenConvAddRelu(std::shared_ptr<Graph>& graph) {
-  fuseFrozenConvAddReluImpl(graph);
+  fuseFrozenConvAddReluGPUImpl(graph);
+  fuseFrozenConvAddReluCPUImpl(graph);
```

Comment:
We want this to compose with the mkldnn fusion pass, right now it does not because we haven't added handling for  `convolution_relu` in that pass. I think the flow of this would be simpler if we just ran a Convolution-Relu pass here after `ComputeSubgraphInMkldnn`. https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp#L1041. We'd be searching for prim::mkldnn_convolution and we would have to check any invariants or handle the case of `relu_`.